src/ucp/proto/proto_common.c

```diff
@@ -392,163 +426,249 @@ ucp_proto_common_get_reg_cost(const ucp_proto_common_init_params_t *params,
+    ucs_memory_type_t recv_mem_type;
+    ucs_linear_func_t perf_multi;
     uint32_t op_attr_mask;
+    ucs_status_t status;
+    size_t frag_size;
+    double frag_ovrh;
 
     /* Remote access implies zero copy on receiver */
     if (params->flags & UCP_PROTO_COMMON_INIT_FLAG_REMOTE_ACCESS) {
         ucs_assert(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY);
     }
 
-   /* TODO
-     * - consider remote/local system device
-     * - consider memory type for pack/unpack
-     */
-    caps->cfg_thresh   = params->cfg_thresh;
-    caps->cfg_priority = params->cfg_priority;
-    caps->min_length   = params->min_length;
-    caps->num_ranges   = 0;
-
-    /* Collect latency and overhead from all lanes */
-    overhead = 0;
-    latency  = params->latency;
-    ucs_for_each_bit(lane, perf_params->lane_map) {
-        iface_attr    = ucp_proto_common_get_iface_attr(&params->super, lane);
-        tl_latency    = ucp_tl_iface_latency(context, &iface_attr->latency);
-        tl_min_length = ucp_proto_common_get_iface_attr_field(
-                iface_attr, params->min_frag_offs, 0);
-
-        overhead        += iface_attr->overhead;
-        latency          = ucs_max(tl_latency, latency);
-        caps->min_length = ucs_max(caps->min_length, tl_min_length);
-    }
+    op_attr_mask = ucp_proto_select_op_attr_from_flags(select_param->op_flags);
 
-    /* Take fragment size from first lane */
-    frag_size = perf_params->frag_size;
-    if (!(params->flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE)) {
-        /* if the data returns as a response, no need to subtract header size */
-        frag_size -= params->hdr_size;
+    /* Calculate sender overhead */
+    if (params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY) {
+        send_ovrh = ucp_proto_common_memreg_time(params, reg_md_map);
+    } else {
+        ucs_assert(reg_md_map == 0);
+        status = ucp_proto_common_buffer_copy_time(
+                params->super.worker, "send-copy", UCS_MEMORY_TYPE_HOST,
+                select_param->mem_type, params->memtype_op, &send_ovrh);
+        if (status != UCS_OK) {
+            return status;
+        }
     }
 
-    op_attr_mask  = ucp_proto_select_op_attr_from_flags(
-                            params->super.select_param->op_flags);
-    uct_time      = ucs_linear_func_make(latency, 1.0 / perf_params->bandwidth);
-    pack_time     = ucs_linear_func_make(0, 1.0 / context->config.ext.bcopy_bw);
-    extra_time    = ucp_proto_common_get_reg_cost(params, perf_params->reg_md_map);
-    extra_time.c += overhead + params->overhead;
+    /* Add constant CPU overhead */
+    send_ovrh.c += perf->overhead;
+
+    /* Add the latency of response/ACK back from the receiver.
+       It's counted as overhead because it does not keep the transport busy. */
+    if ((params->flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE) ||
+        ((op_attr_mask & UCP_OP_ATTR_FLAG_FAST_CMPL) &&
+         (params->flags & UCP_PROTO_COMMON_INIT_FLAG_SEND_ZCOPY))) {
```

Comment:
why is it relevant for zcopy with fast completion only?

Reply:
how is it different without `FAST_CMPL`

Reply:
with FAST_CMPL, we measure the time to finish send request. In case of Zcopy/response, it also waits for ACK/reply.
without FAST_CMPL, we measure how long it takes the data to arrive to the remote side only. So no need to count the ACK back from reciever to sender.

Reply:
ok, ic